File: src/main/scala/leetcode/p973/Solution.scala
package scala.leetcode.p973

import scala.collection.{SortedMap, mutable}


object Blub extends App{

  val v = Array(1,2,3,4,5)

  /*
  for (int i = 0; i < v.size; i++){
    v(i) = v(i)*2;
  */

  for (i <- 0 until v.length/2){
    //v(i) = v(i) *2;
    println("...")
  }

  val res = v.map(x => x*2)

  val totalres = res.reduce( (a,b) => a + b)

  case class Signal(s: String)
  def remainingCharsCount():Int = 15

  val res1 = if (remainingCharsCount() >= 15) Signal("green")
  else if (remainingCharsCount() < 0) Signal("red")
  else Signal("orange")

  val res2 = Signal {
    val remaining = remainingCharsCount()
    if (remaining >= 15) {
      "green"
    } else if (remaining >= 0) {
      "orange"
    } else {
      "red"
    }
  }

  println(res1)
  println(res2)

}


object Solution {

  System.nanoTime()

  def kClosest(points: Array[Array[Int]], K: Int): Array[Array[Int]] = {
    val res = points.map(point => {
      val x = point(0)
      val y = point(1)
      val distance = Math.sqrt(x*x + y*y)
      distance -> point
    })

    implicit val ord:Ordering[Array[(Double,Array[(Int)])]] = ???
    val pq = mutable.PriorityQueue(res)
    pq.drop(K-1)
    ???
  }
}

File: java/src/leetcode/p973/Solution.java
package leetcode.p973;



import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;

public class Solution {

    class Entry{
        double distance;
        int [] point;
        Entry(double distance, int[] point) {
            this.distance = distance;
            this.point = point;
        }

    }

    public static void main(String[] varg){
        int[][] points =  {{1,3},{-2,2}};
        int K = 1;
        Solution sol = new Solution();
        sol.kClosest(points,K);
    }

    public Entry createEntry(int [] point){
        int x= point[0];
        int y = point[1];
        double distance = Math.sqrt(x*x + y*y);
        return new Entry(distance,point);
    }

    public int[][] kClosest(int[][] points, int K) {

        PriorityQueue<Entry> pq = new PriorityQueue<>(new Comparator<Entry>() {
            @Override
            public int compare(Entry o1, Entry o2) {
                if (o1.distance < o2.distance)
                    return -1;
                else if (o1.distance > o2.distance)
                    return 1;
                else
                return 0;
            }
        });
        for (int i = 0; i< points.length; i++){
            pq.add(createEntry(points[i]));
        }


        int[][] result = new int[K][];

        for (int i = 0; i < K; i++){
            result[i] = pq.remove().point;
        }


        return result;
    }

}
